Compiler/src/ssair/EscapeAnalysis.jl

```diff
@@ -225,59 +328,114 @@ end
-    if isa(xa, Bool)
-        xa && ya !== true && return false
-    elseif isa(xa, IndexableFields)
-        if isa(ya, IndexableFields)
-            xinfos, yinfos = xa.infos, ya.infos
-            xn, yn = length(xinfos), length(yinfos)
-            xn > yn && return false
+    x.ThrownEscape ≤ y.ThrownEscape || return false
+    x.ObjectInfo ⊑ₒ y.ObjectInfo || return false
+    x.Liveness ⊑ₗ y.Liveness || return false
+    return true
+end
+
+x::Liveness ⊑ₗ y::Liveness = begin
+    @nospecialize
+    if x === ⊥ₗ
+        return true
+    elseif x === ⊤ₗ
+        return y === ⊤ₗ
+    elseif y === ⊥ₗ
+        return false
+    elseif y === ⊤ₗ
+        return true
+    else
+        x, y = x::PCLiveness, y::PCLiveness
+        return x.pcs ⊆ y.pcs
+    end
+end
+
+x::ObjectInfo ⊑ₒ y::ObjectInfo = begin
+    @nospecialize
+    if x === ⊥ₒ
+        return true
+    elseif x === ⊤ₒ
+        return y === ⊤ₒ
+    elseif y === ⊥ₒ
+        return false
+    elseif y === ⊤ₒ
+        return true
+    elseif x isa HasIndexableCallerFields
+        if y isa HasIndexableCallerFields
+            xcallers, ycallers = x.caller_memory_list, y.caller_memory_list
+            xn, yn = length(xcallers), length(ycallers)
+            xn ≤ yn || return false
             for i in 1:xn
-                xinfos[i] ⊆ yinfos[i] || return false
-            end
-        elseif isa(ya, Unindexable)
-            xinfos, yinfo = xa.infos, ya.info
-            for i = length(xinfos)
-                xinfos[i] ⊆ yinfo || return false
+                xcallerᵢ = xcallers[i]
+                ycallerᵢ = ycallers[i]
+                if !(xcallerᵢ isa IdSet{CallerMemory})
+                    if !(ycallerᵢ isa IdSet{CallerMemory})
+                        xcallerᵢ == ycallerᵢ || return false
+                    else
+                        xcallerᵢ ∈ ycallerᵢ || return false
+                    end
+                else
+                    if !(ycallerᵢ isa IdSet{CallerMemory})
+                        return false
+                    else
+                        xcallerᵢ ⊆ ycallerᵢ || return false
+                    end
+                end
             end
+            xfields, yfields = x.fields, y.fields
+            @goto compare_xfields_yfields
         else
-            ya === true || return false
+            return false
         end
     else
-        xa = xa::Unindexable
-        if isa(ya, Unindexable)
-            xinfo, yinfo = xa.info, ya.info
-            xinfo ⊆ yinfo || return false
+        x = x::HasIndexableFields
+        if y isa HasIndexableCallerFields
+            xfields, yfields = x.fields, y.fields
+            @goto compare_xfields_yfields
         else
-            ya === true || return false
+            y = y::HasIndexableFields
+            xfields, yfields = x.fields, y.fields
+            @label compare_xfields_yfields
+            xn, yn = length(xfields), length(yfields)
+            xn ≤ yn || return false
```

Comment:
Should this be `≤` or `==`?

Does x having a different number of indexable fields imply that x is a different (incomparable) set of alias information?

Reply:
Ah, scratch this - our merge means that we need to flow information from fields indexed from distinct types